lib/python/rose/config_diff.py

```diff
@@ -0,0 +1,230 @@
+    if only_these_options is None:
+        # Default to every option.
+        only_these_options = sorted(metadata.keys())
+    for property_ in only_these_options:
+        value = metadata.get(property_)
+        if value is None:
+            continue
+        id_node.set([property_], value=value)
+    string_file = StringIO.StringIO()
+    rose.config.dump(id_node, target=string_file)
+    return string_file.getvalue()
+
+
+def main():
+    """Implement the "rose config-diff" command."""
+    opt_parser = rose.opt_parse.RoseOptionParser()
+    opt_parser.add_my_options("diff_tool", "graphical",
+                              "ignore", "meta_path", "properties")
+    my_sys_args = list(sys.argv)
+
+    opts, args = opt_parser.parse_args(my_sys_args[1:])
+    rose.macro.add_site_meta_paths()
+    rose.macro.add_env_meta_paths()
+    rose.macro.add_opt_meta_paths(opts.meta_path)
+
+    paths, diff_args = args[:2], args[2:]
+
+    if len(paths) != 2:
+        sys.exit(opt_parser.get_usage())
+
+    config_loader = rose.config.ConfigLoader()
+
+    if opts.properties is None:
+        properties = _DEFAULTS.PROPERTIES.split(",")
+    else:
+        properties = opts.properties.split(",")
+
+    ignore_regexes = expand_shorthand_in_ignore_regexes(
+        opts.ignore_patterns)
+
+    output_filenames = []
+    config_type = rose.SUB_CONFIG_NAME
+    for path in paths:
+        if path == "-":
+            continue
+        config_type = os.path.basename(path)
+    for path in paths:
+        if path == "-":
+            path = sys.stdin
+            filename = config_type
+            directory = None
+        else:
+            directory, filename = path.rsplit(os.sep, 1)
+        config = config_loader.load_with_opts(path)
+        meta_config_tree = rose.macro.load_meta_config_tree(
+            config, directory=directory, config_type=filename)
+        if meta_config_tree is None:
+            meta_config = rose.config.ConfigNode()
+        else:
+            meta_config = meta_config_tree.node
+        annotated_config = annotate_config_with_metadata(
+            config, meta_config, ignore_regexes=ignore_regexes,
+            metadata_properties=properties
+        )
+        output_dir = tempfile.mkdtemp()
+        output_path = os.path.join(output_dir, filename)
+        rose.config.dump(annotated_config, target=output_path)
+        output_filenames.append(output_path)
+    popener = rose.popen.RosePopener()
+    cmd_opts_args = diff_args + output_filenames
+    if opts.diff_tool is None:
+        if opts.graphical_mode:
+            diff_cmd = popener.get_cmd("gdiff_tool", *cmd_opts_args)
+        else:
+            diff_cmd = popener.get_cmd("diff_tool", *cmd_opts_args)
+    else:
+        diff_cmd = shlex.split(opts.diff_tool) + cmd_opts_args
+    return_code = 1
+    try:
+        return_code, stdout, stderr = popener.run(*diff_cmd)
+        sys.stdout.write(stdout)
+        sys.stderr.write(stderr)
```

Comment:
It is still worth doing this:

```python
return_code, stdout, stderr = popener.run(*diff_cmd, stdout=sys.stdout, stderr=sys.stderr)
```

This allows `subprocess.Popen` to write the outputs as soon as they appear.